### PR TITLE
[QUOTA] Chunk by chunk accounting of quota usage.

### PR DESCRIPTION
**Pull request by ankagrawal**
This PR defines a `QuotaChargeEventListener` interface which is passed to frontend router's request handlers to charge for each chunk being served for download or upload.

### CODE REVIEW

File: ambry-api/src/main/java/com/github/ambry/router/Router.java
Comment (justinlin-linkedin) [inline ambry-api/src/main/java/com/github/ambry/router/Router.java:42]: I think we should create a different method for this one, since there are some other callers who don't want to use a eventListener here.

File: ambry-api/src/main/java/com/github/ambry/router/Router.java
Comment (justinlin-linkedin) [inline ambry-api/src/main/java/com/github/ambry/router/Router.java:43]: And can we rename this to something else, like `ChunkEventListener`. I think it can be a generic callback not just for quota charging logic.

File: ambry-frontend/src/main/java/com/github/ambry/frontend/FrontendRestRequestService.java
Comment (justinlin-linkedin) [inline ambry-frontend/src/main/java/com/github/ambry/frontend/FrontendRestRequestService.java:784]: we probably don't need to do Chunk by Chunk on Delete.

File: ambry-api/src/main/java/com/github/ambry/router/Router.java
Comment (justinlin-linkedin) [inline ambry-api/src/main/java/com/github/ambry/router/Router.java:74]: BTW, I don't think stitchBlob requires a callback here. All the chunk data has already been saved in ambry when we call stitchBlob.

File: ambry-frontend/src/main/java/com/github/ambry/frontend/UserQuotaRequestCostPolicy.java
Comment (justinlin-linkedin) [inline ambry-frontend/src/main/java/com/github/ambry/frontend/UserQuotaRequestCostPolicy.java:93]: There is no need to remove this logic since the request cost is also used in C2S events.

File: ambry-frontend/src/test/java/com/github/ambry/frontend/FrontendRestRequestServiceTest.java
Comment (justinlin-linkedin) [inline ambry-frontend/src/test/java/com/github/ambry/frontend/FrontendRestRequestServiceTest.java:138]: we should definitely add new method in Router interface to avoid these empty callbacks.

File: ambry-quota/src/main/java/com/github/ambry/quota/capacityunit/CapacityUnitThrottlePolicy.java
Comment (justinlin-linkedin) [inline ambry-quota/src/main/java/com/github/ambry/quota/capacityunit/CapacityUnitThrottlePolicy.java:3]: what this?

File: ambry-router/src/main/java/com/github/ambry/router/PutOperation.java
Comment (justinlin-linkedin) [inline ambry-router/src/main/java/com/github/ambry/router/PutOperation.java:159]: hmm, can we not move the code, it messes up the review.

File: ambry-api/src/main/java/com/github/ambry/quota/QuotaChargeEventListener.java
Comment (justinlin-linkedin) [inline ambry-api/src/main/java/com/github/ambry/quota/QuotaChargeEventListener.java:42]: I think we can change this to ChunkEventListener and passes some info this listener, like chunkId, chunkType, chunkSize, encryption , etc...

File: ambry-api/src/main/java/com/github/ambry/router/Router.java
Comment (ankagrawal) [reply ambry-api/src/main/java/com/github/ambry/router/Router.java:42]: We can either create a different method, or pass null or empty lambda when callers don't want to use it. Keeping just one method helps keep Router less complicated in my opinion.
Still if you think otherwise, I can fix it.

File: ambry-api/src/main/java/com/github/ambry/router/Router.java
Comment (ankagrawal) [reply ambry-api/src/main/java/com/github/ambry/router/Router.java:43]: The reason why I wanted to keep it as quota specific, instead of a generic interface, is that its not necessary that quota charge will happen on chunks. For example, in case of tll, delete and undelete, we are doing quota charge on the whole operation, and not on individual chunks. So I pass the `QuotaChargeEventListener` to each type `Operation` (e.g, `GetOperation`, `DeleteOperation` etc), and let the operation class itself decide where the quota charge needs to happen (on completion of main operation, or for each chunk). Let me know what you think about this.

The other alternative is to do quota charging at different places for different type of Ambry requests. I feel that this option scatters the quota code everywhere in the stack and something uniform is better.

File: ambry-frontend/src/main/java/com/github/ambry/frontend/FrontendRestRequestService.java
Comment (ankagrawal) [reply ambry-frontend/src/main/java/com/github/ambry/frontend/FrontendRestRequestService.java]: Although we are passing `QuotaChargeEventListener` to all methods of `Router`, we are not charging quota on chunk operation for every operation. In delete, ttlupdate and undelete operation, we are charging quota when the main blob operation is complete.

File: ambry-api/src/main/java/com/github/ambry/router/Router.java
Comment (ankagrawal) [reply ambry-api/src/main/java/com/github/ambry/router/Router.java:74]: stitch blob is charged for only one chunk. Since we create and save the metadata chunk in the stitch blob call, we charge one unit of quota.

File: ambry-frontend/src/main/java/com/github/ambry/frontend/UserQuotaRequestCostPolicy.java
Comment (ankagrawal) [reply ambry-frontend/src/main/java/com/github/ambry/frontend/UserQuotaRequestCostPolicy.java]: fixed.

File: ambry-quota/src/main/java/com/github/ambry/quota/capacityunit/CapacityUnitThrottlePolicy.java
Comment (ankagrawal) [reply ambry-quota/src/main/java/com/github/ambry/quota/capacityunit/CapacityUnitThrottlePolicy.java]: removed.

**Review by justinlin-linkedin — CHANGES_REQUESTED** [review]

File: ambry-router/src/main/java/com/github/ambry/router/PutOperation.java
Comment (justinlin-linkedin) [inline ambry-router/src/main/java/com/github/ambry/router/PutOperation.java:1310]: We have to set the exception fo this PutOperation here in the catch statement, otherwise, it's not consistent with GetBlobOperation since we are setting exception in GetBlobOperation.

File: ambry-api/src/main/java/com/github/ambry/config/QuotaConfig.java
Comment (cgtz) [inline ambry-api/src/main/java/com/github/ambry/config/QuotaConfig.java:82]: "Should requests in progress be throttled if they exceed their quota"

File: ambry-api/src/main/java/com/github/ambry/quota/QuotaChargeEventListener.java
Comment (cgtz) [reply ambry-api/src/main/java/com/github/ambry/quota/QuotaChargeEventListener.java:59]: I feel like the chunk size at the very least may be useful, even if we keep this quota specific. It is a bit of an unknown if the current strategy of treating each chunk as one CU will be granular enough, so maybe providing the size of the chunk may be useful for logging/fractional CU in the future.

File: ambry-api/src/main/java/com/github/ambry/quota/QuotaEnforcer.java
Comment (cgtz) [inline ambry-api/src/main/java/com/github/ambry/quota/QuotaEnforcer.java:44]: How does this method know which type of quota to charge against (usually given by QuotaName)?

File: ambry-api/src/main/java/com/github/ambry/router/Router.java
Comment (cgtz) [reply ambry-api/src/main/java/com/github/ambry/router/Router.java:42]: Have we considered adding `quotaChargeEventListener` as a `GetBlobOption`/`PutBlobOption`? These have associated builders that can help with optional arguments like this.

We could create options classes for delete/undelete/ttlupdate to make it symmetric. Alternatively, do those methods even need to be passed a listener as opposed to the frontend handling this logic in the callback.

File: ambry-api/src/main/java/com/github/ambry/quota/QuotaEnforcer.java
Comment (ankagrawal) [reply ambry-api/src/main/java/com/github/ambry/quota/QuotaEnforcer.java:44]: It looks at RestRequest's `getRestMethod` and converts it to write or read operation. For writes, we can charge for storage quota and WCU. For reads, we only charge for RCU.

File: ambry-api/src/main/java/com/github/ambry/quota/QuotaResource.java
Comment (ankagrawal) [reply ambry-api/src/main/java/com/github/ambry/quota/QuotaResource.java:84]: Its used in closed source code.

File: ambry-quota/src/main/java/com/github/ambry/quota/storage/StorageQuotaEnforcer.java
Comment (ankagrawal) [reply ambry-quota/src/main/java/com/github/ambry/quota/storage/StorageQuotaEnforcer.java:54]: Yes this will need to be revised.

File: ambry-router/src/main/java/com/github/ambry/router/PutOperation.java
Comment (ankagrawal) [reply ambry-router/src/main/java/com/github/ambry/router/PutOperation.java:1310]: Removed setting exception in GetBlobOperation as we are not throttling for in-progress requests for now.

**Review by ankagrawal — COMMENTED** [review]
Initial response to review comments. Code changes for majority of the review comments coming up next.

File: ambry-quota/src/main/java/com/github/ambry/quota/storage/StorageQuotaEnforcer.java
Comment (ankagrawal) [reply ambry-quota/src/main/java/com/github/ambry/quota/storage/StorageQuotaEnforcer.java:54]: fixed.

File: ambry-quota/src/main/java/com/github/ambry/quota/storage/StorageQuotaEnforcer.java
Comment (ankagrawal) [reply ambry-quota/src/main/java/com/github/ambry/quota/storage/StorageQuotaEnforcer.java:131]: done.

File: ambry-quota/src/main/java/com/github/ambry/quota/storage/StorageQuotaEnforcer.java
Comment (ankagrawal) [reply ambry-quota/src/main/java/com/github/ambry/quota/storage/StorageQuotaEnforcer.java:316]: removed.

File: ambry-router/src/main/java/com/github/ambry/router/NonBlockingRouter.java
Comment (ankagrawal) [reply ambry-router/src/main/java/com/github/ambry/router/NonBlockingRouter.java]: fixed.

File: ambry-router/src/main/java/com/github/ambry/router/NonBlockingRouter.java
Comment (ankagrawal) [reply ambry-router/src/main/java/com/github/ambry/router/NonBlockingRouter.java]: fixed.

File: ambry-api/src/main/java/com/github/ambry/router/Router.java
Comment (ankagrawal) [reply ambry-api/src/main/java/com/github/ambry/router/Router.java:42]: Putting the listener or callback in Options class for each operation will make the APIs cleaner. We can take it as a follow up item after this PR is merged.

File: ambry-frontend/src/test/java/com/github/ambry/frontend/FrontendRestRequestServiceTest.java
Comment (ankagrawal) [reply ambry-frontend/src/test/java/com/github/ambry/frontend/FrontendRestRequestServiceTest.java]: Using null now.

File: ambry-api/src/main/java/com/github/ambry/quota/QuotaChargeEventListener.java
Comment (ankagrawal) [reply ambry-api/src/main/java/com/github/ambry/quota/QuotaChargeEventListener.java]: I will take this up as a follow up item after this PR is merged.

File: ambry-frontend/src/main/java/com/github/ambry/frontend/NamedBlobPutHandler.java
Comment (ankagrawal) [reply ambry-frontend/src/main/java/com/github/ambry/frontend/NamedBlobPutHandler.java]: fixed.

File: ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java
Comment (ankagrawal) [reply ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java]: done.

File: ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java
Comment (ankagrawal) [reply ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java]: fixed.

File: ambry-router/src/main/java/com/github/ambry/router/NonBlockingRouter.java
Comment (ankagrawal) [reply ambry-router/src/main/java/com/github/ambry/router/NonBlockingRouter.java]: We are delegating the quota charge callback responsibility to Operation classes. They can handle how and when they want to charge for quota.

File: ambry-router/src/main/java/com/github/ambry/router/NonBlockingRouter.java
Comment (ankagrawal) [reply ambry-router/src/main/java/com/github/ambry/router/NonBlockingRouter.java]: The implementation of `onQuotaChargeEvent` makes sure that the only exception thrown out of that method will be RouterException.

File: ambry-quota/src/main/java/com/github/ambry/quota/AmbryQuotaManager.java
Comment (justinlin-linkedin) [inline ambry-quota/src/main/java/com/github/ambry/quota/AmbryQuotaManager.java:125]: This method should probably stay for future usage.

File: ambry-router/src/main/java/com/github/ambry/router/BatchOperationCallbackTracker.java
Comment (justinlin-linkedin) [inline ambry-router/src/main/java/com/github/ambry/router/BatchOperationCallbackTracker.java:49]: why do we need another callback here, we already have `callback`, there is no need to have `completionCallback`.

File: ambry-router/src/main/java/com/github/ambry/router/BatchOperationCallbackTracker.java
Comment (justinlin-linkedin) [inline ambry-router/src/main/java/com/github/ambry/router/BatchOperationCallbackTracker.java:89]: The callback will be invoked in line 91. we don't need another one.

File: ambry-router/src/main/java/com/github/ambry/router/GetBlobInfoOperation.java
Comment (justinlin-linkedin) [inline ambry-router/src/main/java/com/github/ambry/router/GetBlobInfoOperation.java:363]: nit: can we do this in `checkAndMaybeComplete` method, just like what we did in `DeleteOperation`.

File: ambry-router/src/main/java/com/github/ambry/router/NonBlockingRouterFactory.java
Comment (justinlin-linkedin) [inline ambry-router/src/main/java/com/github/ambry/router/NonBlockingRouterFactory.java:127]: what is this?

File: ambry-router/src/main/java/com/github/ambry/router/TtlUpdateOperation.java
Comment (justinlin-linkedin) [inline ambry-router/src/main/java/com/github/ambry/router/TtlUpdateOperation.java:340]: we are charging too many for TTLUPDATE and UNDELETE. 
We should only charge once for TTL_UPDATE and UNDELETE, and we already checked it on TtlUpdateManager and UndeleteManager, we don't need to change it again in TtlUpdateOperation.

File: ambry-router/src/test/java/com/github/ambry/router/NonBlockingRouterQuotaCallbackTest.java
Comment (justinlin-linkedin) [inline ambry-router/src/test/java/com/github/ambry/router/NonBlockingRouterQuotaCallbackTest.java:41]: I think taking QuotaCallback test out from NonBlockingRouterTest is unnecessary. We only have one test method and we are take huge chunk of code and create a base file. It's too many changes for too little gain.

File: ambry-quota/src/main/java/com/github/ambry/quota/AmbryQuotaManager.java
Comment (ankagrawal) [reply ambry-quota/src/main/java/com/github/ambry/quota/AmbryQuotaManager.java:125]: The main implementation for this interface is in closed source code. The method exists there.

File: ambry-router/src/main/java/com/github/ambry/router/NonBlockingRouterFactory.java
Comment (ankagrawal) [reply ambry-router/src/main/java/com/github/ambry/router/NonBlockingRouterFactory.java:127]: Not sure why its here. Maybe due to rebase.

File: ambry-router/src/main/java/com/github/ambry/router/TtlUpdateOperation.java
Comment (ankagrawal) [reply ambry-router/src/main/java/com/github/ambry/router/TtlUpdateOperation.java:340]: ttlupdate and undelete have different flows for a simple blob and for a composite blob. This charge here will happen only for the case of a simple blob's ttl update. The reason being that for the BatchOperation, the quotaChargeCallback is passed as null. Hence for composite blobs, only one charge will happen when all the blobs are updated.

File: ambry-router/src/test/java/com/github/ambry/router/NonBlockingRouterQuotaCallbackTest.java
Comment (ankagrawal) [reply ambry-router/src/test/java/com/github/ambry/router/NonBlockingRouterQuotaCallbackTest.java:41]: The main motivation behind taking this out was that I want to test it for a much large number of parameters. And I don't want to test all the test cases of original NonBlockingRouterTest for all those parameters. This slows down our test time for the build.

File: ambry-router/src/main/java/com/github/ambry/router/BatchOperationCallbackTracker.java
Comment (ankagrawal) [reply ambry-router/src/main/java/com/github/ambry/router/BatchOperationCallbackTracker.java]: fixed.

Comment (ankagrawal) [reply ambry-router/src/main/java/com/github/ambry/router/BatchOperationCallbackTracker.java]: fixed.

File: ambry-router/src/main/java/com/github/ambry/router/BatchOperationCallbackTracker.java
Comment (ankagrawal) [reply ambry-router/src/main/java/com/github/ambry/router/BatchOperationCallbackTracker.java]: fixed.

File: ambry-router/src/main/java/com/github/ambry/router/BatchOperationCallbackTracker.java
Comment (justinlin-linkedin) [inline ambry-router/src/main/java/com/github/ambry/router/BatchOperationCallbackTracker.java:94]: we should move this blob inside of the if blob below, like
```
    if (completed.compareAndSet(false, true)) {
        if(quotaChargeCallback != null) {
            try {
                quotaChargeCallback.chargeQuota();
            } catch (RouterException rEx) {
                LOGGER.info("Exception {} while charging quota for ttl operation", rEx.toString());
            }
      }
      NonBlockingRouter.completeOperation(futureResult, callback, null, e, false);
    }
```

File: ambry-router/src/main/java/com/github/ambry/router/NonBlockingRouterFactory.java
Comment (justinlin-linkedin) [inline ambry-router/src/main/java/com/github/ambry/router/NonBlockingRouterFactory.java:18]: These imports don't get used anywhere.

File: ambry-router/src/main/java/com/github/ambry/router/TtlUpdateOperation.java
Comment (justinlin-linkedin) [reply ambry-router/src/main/java/com/github/ambry/router/TtlUpdateOperation.java:340]: that makes sense

**Review by justinlin-linkedin — APPROVED** [review]

File: ambry-router/src/main/java/com/github/ambry/router/NonBlockingRouterFactory.java
Comment (ankagrawal) [reply ambry-router/src/main/java/com/github/ambry/router/NonBlockingRouterFactory.java]: fixed.

File: ambry-router/src/main/java/com/github/ambry/router/BatchOperationCallbackTracker.java
Comment (ankagrawal) [reply ambry-router/src/main/java/com/github/ambry/router/BatchOperationCallbackTracker.java]: fixed.

File: ambry-router/src/main/java/com/github/ambry/router/DeleteOperation.java
Comment (cgtz) [inline ambry-router/src/main/java/com/github/ambry/router/DeleteOperation.java:312]: I think there is an issue here. `checkAndMaybeComplete` is intended as an idempotent method that can be called multiple times. For example, let's look at the `handleResponse`  call. This is called for every response that comes back from a server. I think it may be possible for a response to come back such that the success threshold for the operation comes back, and this code block will be executed. Then, the 3rd successful response may come back and this would be called again.

File: ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java
Comment (cgtz) [inline ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java:851]: For the actions that are done from within checkAndMaybeComplete in all classes, I would suggest that we call the callback only if the completion status is not already ChunkState.Complete:

```
if (state != ChunkState.Complete && quotaChargeCallback != null) { chargeQuota }
```

This comment applies to all operation classes

File: ambry-router/src/main/java/com/github/ambry/router/GetBlobInfoOperation.java
Comment (cgtz) [reply ambry-router/src/main/java/com/github/ambry/router/GetBlobInfoOperation.java:363]: +1 let's centralize on checkAndMaybeComplete with extra state checking logic to ensure the callback is called only once.

File: ambry-router/src/main/java/com/github/ambry/router/UndeleteOperation.java
Comment (cgtz) [inline ambry-router/src/main/java/com/github/ambry/router/UndeleteOperation.java:365]: all of these code blocks seem to be missing a space in between if and paren.

**Review by cgtz — APPROVED** [review]
LGTM after fixing issues to make sure callback only called once.

File: ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java
Comment (ankagrawal) [reply ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java:851]: done.

File: ambry-router/src/main/java/com/github/ambry/router/UndeleteOperation.java
Comment (ankagrawal) [reply ambry-router/src/main/java/com/github/ambry/router/UndeleteOperation.java:365]: fixed.

**Review by cgtz — APPROVED** [review]

**Pull request merged**
